include/ttmlir/Dialect/TTKernel/IR/TTKernelOps.td

```diff
@@ -438,6 +438,136 @@ def TTKernel_NocAsyncWriteBarrierOp : TTKernel_Op<"noc_async_write_barrier"> {
     }];
 }
 
+def TTKernel_GetSemaphoreOp : TTKernel_Op<"get_semaphore"> {
+    let summary = "GetSemaphoreOp";
+    let description = [{
+      Get L1 addr of the semaphore with specified semaphore id
+    }];
+
+    let arguments = (ins I32:$semaphore_id);
```

Comment:
Should this not be of `TTKernel_Semaphore` type? emitC can lower the Semaphore type to an int.  I feel like we can make some things slightly better typed than their corresponding dataflow API

Reply:
Makes sense, I have tried something similar, but what I found as a problem is the chaining of different ops. For example, `GetArgValOp` is expected to return `I32` type because we won't always use it to get semaphore address. Because of that, when we get `I32` from `GetArgValOp`, that is what we are using as input to other functions, like `get_semaphore`.  We could possibly have another op, something like `GetSemaphoreIdOp`, that would be lowered to `get_arg_val` as well, but I am not sure whether that is something that we want to do in order to improved types in the ops. Let me know what you think

Reply:
As a helper to this question, can I somehow derive Types further from already existing types. For example derive new type from `TTKernel_L1Addr`? https://github.com/tenstorrent/tt-mlir/pull/919/files#diff-65d83242f558f0415bf13b65d787b1060f0ff1880e46cf3581f9e7e6f2e43782R120

Reply:
We could have the result type of `GetArgValOp` be `AnyTypeOf<[Semaphore, I32]>` so that it could have stronger typing. Lmk if you think this sounds like overkill though.

Reply:
> As a helper to this question, can I somehow derive Types further from already existing types? For example derive new type from [`TTKernel_L1Addr`](https://github.com/tenstorrent/tt-mlir/pull/919/files#diff-65d83242f558f0415bf13b65d787b1060f0ff1880e46cf3581f9e7e6f2e43782R120)

I think the closest thing you could do is make a `class` in tablegen that provides some kind of common interface, but I'm not sure it's worth it.  I don't think it's possible to inherit.